dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,187 @@
+# *****************************************************************************
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# ******************************************************************************
+import requests
+import urllib3
+
+from dlab_core.domain.exceptions import DLabException
+from dlab_core.domain.helper import validate_property_type
+
+from jose import jwt
+
+
+URL_INTROSPECT = '{realm_address}/protocol/openid-connect/token/introspect'
+
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+
+class KeyCloakException(DLabException):
+    pass
+
+
+class KeyCloakConnectionException(KeyCloakException):
+    pass
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        if not (isinstance(keycloak_host, str) and isinstance(realm_name, str)):
```

Comment:
why don't you use properties with validate decorator for these cases?

Reply:
I did not want to make a lot of properties with similar code, so I ended up with using descriptor for that:
`
class StringProperty(object):
    """A descriptor that forbids non-string values """

    def __init__(self, value=None):
        self.value = value

    def __get__(self, instance, owner):
        return self.value

    @validate_property_type(str, arg_index=2)
    def __set__(self, instance, value):
        self.value = value
`